engine/tests/phpunit/integration/Elgg/Integration/ElggEntityMetadataTest.php

```diff
@@ -0,0 +1,285 @@
+			'metadata_name' => 'foo',
+		]);
+		$this->assertCount(1, $md);
+		$new_id = $md[0]->id;
+		
+		$this->entity->foo = 'bar2';
+		$md = elgg_get_metadata([
+			'guid' => $this->entity->guid,
+			'metadata_name' => 'foo',
+		]);
+		$this->assertCount(1, $md);
+		$update_id = $md[0]->id;
+		
+		$this->assertNotEquals($new_id, $update_id);
+	}
+		
+	public function testDeleteMetadataNonExistingMetadata() {
+		// let's delete a non-existent metadata
+		$this->assertNull($this->entity->deleteMetadata('non-existing'));
+	}
+		
+	public function testDeleteMetadataNonExistingMetadataNotDeletingOtherMetadata() {
+		// @link https://github.com/elgg/elgg/issues/2273
+		$this->assertTrue($this->entity->setMetadata('foo', 'bar'));
+		$this->assertNull($this->entity->deleteMetadata('notfoo'));
+		$this->assertEquals($this->entity->getMetadata('foo'), 'bar');
+	}
+
+	public function testDeleteMetadataSingleName() {
+		$this->assertTrue($this->entity->setMetadata('foo', 'bar'));
+		$this->assertEquals($this->entity->getMetadata('foo'), 'bar');
+		$this->assertTrue($this->entity->setMetadata('important', 'indeed!'));
+		$this->assertEquals($this->entity->getMetadata('important'), 'indeed!');
+		$this->assertTrue($this->entity->deleteMetadata('important'));
+		$this->assertNull($this->entity->getMetadata('important'));
+		$this->assertEquals($this->entity->getMetadata('foo'), 'bar');
+	}
+
+	public function testDeleteMetadataDeletesAll() {
+		$initial_count = (int) elgg_get_metadata([
+			'guid' => $this->entity->guid,
+			'count' => true,
+		]);
+		
+		$this->entity->foo = 'bar';
+		$this->entity->bar = 'foo';
+		
+		// count if all metadata is deleted
+		$metadata_count = (int) elgg_get_metadata([
+			'guid' => $this->entity->guid,
+			'count' => true,
+		]);
+		$this->assertEquals($initial_count + 2, $metadata_count);
+		
+		$this->entity->deleteMetadata();
+		
+		$this->assertNull($this->entity->foo);
+		$this->assertNull($this->entity->bar);
+		
+		// count if all metadata is deleted
+		$metadata_count = (int) elgg_get_metadata([
+			'guid' => $this->entity->guid,
+			'count' => true,
+		]);
+		$this->assertEquals(0, $metadata_count);
+	}
+	
+	public function testSetMetadataAppendSingleValueReturnsArray() {
+		foreach ($this->entities as $entity) {
+			$entity->foo = 'test1';
+			$entity->setMetadata('foo', 'test2', '', true);
+
+			$this->assertEquals([
+				'test1',
+				'test2'
+			], $entity->foo);
+		}
+	}
+
+	public function testSetMetadataAppendSingleElementArrayValueReturnsArray() {
+		foreach ($this->entities as $entity) {
+			$entity->foo = 'test1';
+			$entity->setMetadata('foo', ['test2'], '', true);
+
+			$this->assertEquals([
+				'test1',
+				'test2'
+			], $entity->foo);
+		}
+	}
+
+	public function testSetMetadataAppendArrayValueReturnsArray() {
+		foreach ($this->entities as $entity) {
+			$entity->foo = [
+				'brett',
+				'bryan',
+				'brad',
+			];
+			$entity->setMetadata('foo', ['test1', 'test2', 'test3'], '', true);
```

Comment:
I don't get this. Why metadata is merged?

Reply:
I suppose it makes sense, but I never realized you could do this. I don't ever use setMetadata, so I guess I never came across this cork, but honestly doesn't feel very intuitive.

Reply:
this was a test that existed and indeed not something that i use also...